### PR TITLE
Complete animations whether or not cascade needs done

### LINKED ISSUE #12554: Width of container with IFrame inside resets when a link in the IFrame is clicked
**Width of container with IFrame inside resets when a link in the IFrame is clicked.  The issue only occurs when the container has a transition on the width property.**

**See test case at: [https://samuknet.github.io/test_cases/iframeWidthReset/](https://samuknet.github.io/test_cases/iframeWidthReset/)**

There is an IFrame in a container 'strip'.  The IFrame is styled to have a width and height of 100% of the container.

When the page loads, JavaScript expands the container width to 100vw.  **The issue occurs whens clicking a link inside the IFrame**.  This causes the parent container to unexpectedly revert to the original width, 'resetting' the width property that had been applied with JavaScript.

The parent container has a CSS transition on the width property so that the initial change of width to 100vw is animated.  Crucially, removing this CSS transition stops the container collapsing when clicking a link in the IFrame, suggesting that the issue is related to the transition property.

### PR DESCRIPTION
**Pull request by notriddle**
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes
- [X] These changes fix #12554 (github issue number if applicable).

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12623)

### CODE REVIEW

File: components/style/traversal.rs [hunk 1]
```diff
@@ -232,6 +234,32 @@ pub fn recalc_style_at<'a, N, C>(context: &'a C,
...
+        {
+            let mut data = node.mutate_data().unwrap();
+            let mut style = data.style.as_mut().unwrap();
+            let all_expired_animations = context.expired_animations.read().unwrap();
+            let animations_to_expire = all_expired_animations.get(&node_opaque);
+            had_animations_to_expire = animations_to_expire.is_some();
+            if let Some(ref animations) = animations_to_expire {
+                for animation in *animations {
+                    // NB: Expiring a keyframes animation is the same as not
+                    // applying the keyframes style to it, so we're safe.
```
Comment (emilio) [inline components/style/traversal.rs]: This is not quite true, could you make a note about supporting [animation fill mode](https://www.w3.org/TR/css3-animations/#animation-fill-mode)?

**Comment by emilio** [comment]
Reviewed 2 of 2 files at r2, 2 of 2 files at r3, 1 of 2 files at r5.
Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/style/matching.rs, line 477 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12623#-KNhzOpsVUkLeuQzFY_w:-KNhzOpsVUkLeuQzFY_x:870258954) ([raw file](https://github.com/servo/servo/blob/6d6c4a553364210a8b5c04bece23c63306a4b16f/components/style/matching.rs#L477)):*
> ```Rust
>                 for animation in *animations {
>                     // TODO: support animation-fill-mode
>                     if let Animation::Transition(_, _, ref frame, _) = *animation {
> ```

Can we factor this code out and use it in both situations?

---

*[components/style/traversal.rs, line 239 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12623#-KNhzVFMsCqmzPZCuDjY:-KNhzVFMsCqmzPZCuDjZ:-1199941283) ([raw file](https://github.com/servo/servo/blob/e8fa0f27a0ca73e294eaeae45d73fc0501660296/components/style/traversal.rs#L239)):*
> ```Rust
>     } else {
>         // Finish any expired transitions, without performing the cascade.
>         let node_opaque = node.opaque();
> ```

I'm kind of worried this is a bit overkill in the common case there aren't transitions to expire. Also, I wonder if this will even work now #12563 has landed, where we won't process the nodes that aren't marked as needing a restyle. We also don't construct flows again for every node, so this might be fixed with that patch?

In case it's not: What about just layout marking nodes as dirty once it detects there's an animation to expire, and running the cascade then? Hmm... But layout only knows about `OpaqueNode`s, and by design these are not convertible back to a node... Maybe they should be from script?

Other way to do it is adding a flag to the node (`IS_ANIMATED`?) in `start_transitions_if_applicable`, but then we should have another flag, let's say `HAS_ANIMATED_CHILDREN`, and keep doing the traversal for these too.

Another solution that may actually work even better is updating the expired animation style from layout (in `recalc_style_for_animations`)?

We could remove the items in the expired animations map there if it's not needed from the cascade (haven't actually tested it), but even if it's needed, updating the style there *should* prevent it getting updated again until a proper restyle with cascade is done.

---

*[components/style/traversal.rs, line 252 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12623#-KNhSxSvIYZOfTNPnwvZ-r1-252:-KNj3DQn6mVI5EzMmNI-:-1834867277) ([raw file](https://github.com/servo/servo/blob/e8fa0f27a0ca73e294eaeae45d73fc0501660296/components/style/traversal.rs#L252)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> This is not quite true, could you make a note about supporting [animation fill mode](https://www.w3.org/TR/css3-animations/#animation-fill-mode)?
</details>
Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12623)*

**Comment by emilio** [comment]
Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/style/traversal.rs, line 239 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12623#-KNhzVFMsCqmzPZCuDjY:-KNj3sg8SMpG3D5jTJA0:-450332821) ([raw file](https://github.com/servo/servo/blob/e8fa0f27a0ca73e294eaeae45d73fc0501660296/components/style/traversal.rs#L239)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> I'm kind of worried this is a bit overkill in the common case there aren't transitions to expire. Also, I wonder if this will even work now #12563 has landed, where we won't process the nodes that aren't marked as needing a restyle. We also don't construct flows again for every node, so this might be fixed with that patch?
> 
> In case it's not: What about just layout marking nodes as dirty once it detects there's an animation to expire, and running the cascade then? Hmm... But layout only knows about `OpaqueNode`s, and by design these are not convertible back to a node... Maybe they should be from script?
> 
> Other way to do it is adding a flag to the node (`IS_ANIMATED`?) in `start_transitions_if_applicable`, but then we should have another flag, let's say `HAS_ANIMATED_CHILDREN`, and keep doing the traversal for these too.
> 
> Another solution that may actually work even better is updating the expired animation style from layout (in `recalc_style_for_animations`)?
> 
> We could remove the items in the expired animations map there if it's not needed from the cascade (haven't actually tested it), but even if it's needed, updating the style there *should* prevent it getting updated again until a proper restyle with cascade is done.
</details>
Actually, thinking about it a bit more, this applies in the case an element isn't restyled, but its children are right? (i.e. the `HAS_DIRTY_DESCENDANTS` flag). In that case, I guess the overhead is not so much.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12623)*

**Comment by notriddle** [comment]
*[components/style/matching.rs, line 477 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12623#-KNhzOpsVUkLeuQzFY_w:-KNj9YYDbMyfDqQuW8Ax:-935591543) ([raw file](https://github.com/servo/servo/blob/6d6c4a553364210a8b5c04bece23c63306a4b16f/components/style/matching.rs#L477)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> Can we factor this code out and use it in both situations?
</details>
I've been thinking about it, and I'm pretty sure I can do that, yeah.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12623)*

File: components/style/traversal.rs [hunk 2]
```diff
@@ -234,33 +234,15 @@ pub fn recalc_style_at<'a, N, C>(context: &'a C,
...
-            context.expired_animations.write().unwrap().remove(&node_opaque);
-        }
     }
 
+    // Finish any expired transitions.
+    animation::complete_expired_transitions(
+        node.opaque(),
+        node.mutate_data().unwrap().style.as_mut().unwrap(),
+        context.shared_context()
+    );
```
Comment (emilio) [inline components/style/traversal.rs]: Aren't we doing this twice now in the cascade case?

**Comment by notriddle** [comment]
*[components/style/traversal.rs, line 244 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12623#-KNhSxSvIYZOfTNPnwvZ-r7-244:-KNjLdREYCH5VkmkJ6bg:-1393162547) ([raw file](https://github.com/servo/servo/blob/3eb5d983ac85c270cb41a6bb8ddd82e070ec27ce/components/style/traversal.rs#L244)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> Aren't we doing this twice now in the cascade case?
</details>
Oh, right, oops.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12623)*

**Comment by emilio** [comment]
@bors-servo: try

Let's check if there's any other transition/animation test that could check this in wpt.

**Comment by notriddle** [comment]
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/mozbrowser/iframe_visibility.html:
  │ FAIL [expected PASS] Requesting animation frame composites only when frame is visible
  │   → assert_true: expected true got false
  │ 
  │ iframe.onload</<@http://web-platform.test:8000/_mozilla/mozilla/mozbrowser/iframe_visibility.html:58:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1402:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1426:20
```

Other than that, no unexpected test results, including passes.

**Comment by emilio** [comment]
Cool, r=me then.

@bors-servo: r+

**Comment by emilio** [comment]
@bors-servo: r-

Could you squash before, please?

@bors-servo: delegate+

**Comment by notriddle** [comment]
Is iframe_visibliity.html a known intermittent? Because there's no issue for it.

**Comment by jdm** [comment]
@notriddle: #11825

**Pull request merged**